_posts/blog/2018-09-13/2018-09-13-how-decentralized-application-development-nearly-destroyed-my-world.markdown

```diff
@@ -0,0 +1,84 @@
+
+![Shrug Emoji](Shrug Emoji.png)
+
+- **Browsers live in a centralized world**
+
+The OpenBazaar front end is developed with Electron, which uses Chromium, the tech behind the Google Chrome browser. As such, it's subject to many of the browser’s restrictions. One of those is that you are limited to 6 concurrent requests to the same domain. Even though we're ultimately getting the data from many nodes, the requests are proxied through each node's server. In other words, from the browser's point of view, a fetch of 20 different user profiles are all coming from the same domain.
+
+If latency weren't an issue this wouldn't be so bad. Even if a request took a less than ideal 250ms, you would be able to fully display a grid of 24 users in one second. But when requests routinely take 30 seconds, 45 seconds or 1 minute plus, suddenly it takes a whole lot of time to fully display a page. Even more problematic is that your request queue can clog up. Let's say one profile comes back in a lightning fast 100ms and we display a user box for that user with a Follow button. If there are still 23 pending requests, even if the user clicks that Follow button, it will be put at the end of the queue and not even be called until all but 5 requests remain, which may easily be a few minutes.
+
+### An Example of How We Addressed These Challenges
+
+OpenBazaar has an integrated chat section that's really an app within an app. The chat app is present from the time your server starts and remains perpetually in the sidebar. Initially, this was a world destroyer because the little **chat heads** (the user icons you click to open a conversation with a user) require some profile data from the party you’re conversing with (e.g. avatar, name). If a user is even slightly chatty and has engaged in at least 6 conversations, then those 6 profile calls would be enough to clog up the request queue and potentially prevent critical app functionality from being available.
+
+![Chat Heads in OpenBazaar](Chat Heads in OpenBazaar.png "Chat Heads in OpenBazaar")
+
+#### Here's how we handled some of the challenges decentralization presented us in OpenBazaar chat:
+
+- **Latency**
+
+The server stores a list of your conversations. Within that data set is the peerId of each person you are chatting with and the last message in the conversation. This allows us to create a chat head with a default avatar and use the peerId as the name. The peerId is also enough to make a link to the user's page. While it may take up to a minute for the user's avatar and name show up, at least you'll have something showing in the client and you won't be prevented from starting the conversation.
```

Comment:
Super minor, but I would spell peerId as peer ID, since readers don't know what the variable name is for. Or put it in a code block like you do for other variables below.

If you change this, remember to change it later in the text when peerId is used.